### PR TITLE
[VUFIND-1701] Setting to display FOLIO holdings info when no items exist

### PR DESCRIPTION
**Pull request by dltj**
> FOLIO allows holdings records with no attached items. This holdings information isn't displayed in the search results, nor on the Holdings tab of the full record screen.

https://openlibraryfoundation.atlassian.net/browse/VUFIND-1701

TODO
- [x] When merging, update changelog to note change to AvailabilityStatusInterface

### CODE REVIEW

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (dltj) [inline module/VuFind/src/VuFind/ILS/Driver/Folio.php]: This assigning of the `'id' => $bibId` is probably the one part I'm not sure about. I found that it was necessary to do this for the AJAX/GetItemStatus::handleRequest() to work at [line 545](https://github.com/vufind-org/vufind/blob/99dec4025b7e3fe3b418e2365bc6b2812dbe3604/module/VuFind/src/VuFind/AjaxHandler/GetItemStatuses.php#L545) where it uses this `id` value to match up against the bib record identifier; that is important because later at line 549 we declare that the record has been "found". But this is technically the holdings record identifier?

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/Folio.php]: You should move this to its own file in the VuFind\ILS\Logic namespace; this isn't FOLIO-specific and might be useful for other drivers, so moving it would make it more reusable. We also have a "one class per file" rule in the style checker, so this move should help get the build passing as well.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/Folio.php]: This comment should probably be changed or removed.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Alternatively, though, I wonder if it would make sense to add a new setter and/or constructor parameter to the existing AvailabilityStatus class to allow visibility toggling/configuration instead of requiring a new subclass at all. I wonder if @ThoWagen has any opinions about this.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/Folio.php]: This lookup is already being done in formatHoldingItem(). I wonder if it would be better to construct a fake item and call that support method instead of building a complete fake array below. (I realize that some refactoring would be required, perhaps including the addition of an optional parameter for complete availability status objects... but I think that would lead to less redundant code and would reduce the odds of things getting out of sync over time). At the very least, if you decide not to refactor, formatHoldingItem() should be revised to use the new location name/code properties instead of a redundant lookup, and your new virtual holding array below should include the code as well as the name for consistency.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @dltj, this is off to a great start. See below for a few comments and suggestions, and please let me know if I can do anything more to help!

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Yes, I think adding a setter and getter in the base class makes sense.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: I struggled with this part of the pull request—the holdings record details themselves are already a kind of fake item in that they are passed to the rest of VuFind as part of any items attached to a bib record. This is the only part where there is overlap between the holdings record fields and the item record fields. Maybe put a comment in each place cross-referencing the two places where the location UUID is translated? I think that would be clearer than trying to force a DRY refactoring here.

I'm not sure about your last comment...the code here is nearly identical to the four lines in `formatHoldingItem()` except that I didn't use a `$locationId` intermediate variable:
```php
        $locationId = $item->effectiveLocation->id;
        $locationData = $this->getLocationData($locationId);
        $locationName = $locationData['name'];
        $locationCode = $locationData['code'];
```

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: I did it this way because the intention of the class as written seemed to be to [override the `isVisibleInHoldings()` method](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php#L89-L98), but I can reconfigure the class if necessary. (In the meantime, I pulled this class out of the Driver and into Logic.)

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Regarding the last part of my comment, I'm referring to the fact that the new 'holdingLocationName' and 'holdingLocationCode' entries in the holding details array could be used in formatHoldingItem now instead of re-fetching the location data.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Since @ThoWagen wrote the original class but agrees with the idea of adding a getter and setter, that seems to be justification to make that change instead of adding a new class. :-)

Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Good enough for me! I'm adding this and removing the new class:
```php
    /**
     * Set visibility in holdings tab.
     */
    public function setVisibility(bool $visibility): AvailabilityStatus
    {
        $this->visibility = $visibility;
        return $this;
    }
```

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: I've stared at this code a bunch. I see the somewhat duplicate code in creating the fake item, but I can't work out a refactoring path that eliminates it.  I'll gladly accept some pointers if you have any to offer.

I'll also accept pointers on updating the tests so that they pass...not sure how the changes I made cause the tests to output`/locations` rather than `/inventory/items-by-holding-id`

File: module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php]: If you set this to true here, then you don't need an extra line in the constructor.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: VuFind doesn't use holdings IDs, it only uses bib IDs and item IDs -- the ILS driver spec flattens everything out. So I think this is probably okay.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: I think my earlier suggestion and the test failure are related to the same thing: the code is now fetching locations twice unnecessarily, instead of reusing the locations it has already fetched. You've added a getLocationData call in line 618, but the method is then getting called a second time in line 680. Since you have access to the results from line 618 via the $holdingDetails array available to you in line 680, I think you should eliminate the second call and reuse the existing stored data. That might help!

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @dltj -- I didn't have time for a comprehensive review today, but I've at least responded to some open threads. :-)

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Oh! I get it now...`getLocationData()` isn't being called twice when there are no items (which is what I was laser-focused on)...it is when there are items attached. I pulled this code out to a helper method. Waiting to see what the tests say.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Minor nitpick: I don't think you need to wrap this line.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Another nitpick: I don't think you need to have VuFind in this method name or related variables. It's all VuFind, after all. ;-) Maybe `getItemFieldsFromLocationAndHoldings` and `$locationAndHoldingsData` would be more descriptive names for method and variable respectively.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/Folio.php]: The 'reserve' value should be capitalized -- so this should be 'N' rather than 'n'.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/Folio.php]: What led to the choice of the values here? `See full record` does not seem to be a string currently included in our language files. Does this actually end up being displayed somewhere? If so, we should make sure the i18n details are in order (either revise to use an existing string, or add a new string to the en.ini file so it can be translated). If it's not displayed anywhere, then maybe we should use different initial values to avoid confusion.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @dltj, this is shaping up very nicely! See below for a few questions and suggestions. (And note that I haven't tried this "hands-on" yet, so apologies if any of my questions are ill-informed). :-)

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: My thought process here was that this was in the driver to distinguish between what is a _VuFind_ data structure and what is an _ILS_ data structure. As a relatively new reader of the ILS driver code, this was a source of confusion, so I wanted to make it explicit here. When refactoring this into a private method, I was also unsure if anything else (beyond location and holdings details) would need to be added here someday.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Will fix!

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Yes, this is displayed on the search results screen. How do I queue up a string for translation? (Do I just add it to the en.ini file and magic will happen elsewhere?)
image

Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Yes, all you need to do is add a line to en.ini. I'd suggest creating a descriptive key instead of using the raw text as the key -- e.g. something like `virtual_item_availability_status_message = "See full record"` so it can be customized more easily and so the translators have a little more context.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: I guess my thought is that we're creating different chunks of data for different aspects of the ultimate return value -- e.g. the call number details, and now the location/holdings details, and then some other context-specific details -- and we're composing them all together into the final return value. This one part that you have separated out into a support method isn't "more VuFind-y" than the other parts, and singling it out as such may introduce a new form of confusion.

Is it possible that things could be clarified through better method comments and/or `@return` declarations -- e.g. if the methods are more clear about "raw FOLIO data in, VuFind-formatted data out" would that address your concerns?

I certainly don't intend to dismiss your concerns here -- I just want to be sure we find the best way to address them! :-)

Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: That makes sense... I made those changes.

File: languages/en.ini
Comment (demiankatz) [inline languages/en.ini]: I forgot that we actually have a separate namespace for these types of messages -- I'd suggest moving this into languages/HoldingStatus/en.ini and then adjusting the string in Folio.php to be "HoldingStatus::holding_no_items_availability_message".

Also note that you can run `php $VUFIND_HOME/public/index language normalize $VUFIND_HOME/languages` to fix the normalization error that's causing a test failure right now.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @dltj! One more minor thing...

File: languages/en.ini
Comment (dltj) [reply languages/en.ini]: Oh, wow! I didn't even know those sub-directories were there!

Okay, let's see if the tests pass with this change. They should...

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @dltj, I think the code is very clear now, and tests are indeed passing!

One more question: should we add a Folio.ini setting to toggle this behavior, in case some people actually want the legacy behavior?

Once that is settled, perhaps we should solicit a couple more reviews from the FOLIO community to be sure nothing has been overlooked, and then I think this can be merged!

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (maccabeelevine) [inline module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Technically this is "if there are no item records on this holdings record".  Often one holdings record per location, but there is nothing mandating that.

File: module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php
Comment (maccabeelevine) [inline module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php]: Does `$visibility` mean the same thing as `visibility in holdings tab`?  It seems like these fake items *are* visible in the search results, in the sense of having a presence there and saying to check availability on the item page.  There might be a more specific name, although I'm not sure `$visibilityInHoldingsTab` is better.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (maccabeelevine) [inline module/VuFind/src/VuFind/ILS/Driver/Folio.php:886]: There an (intentional, I think) inconsistency here where normal items are sent through `$this->formatHoldingItem()` and that output added to `$nextBatch`, whereas for the generated/fake item there is no format step.  It makes sense to not 'format' something that will not be visually displayed.  But the inconsistency bugs me, maybe because there could be use cases (not covered here) for actually displaying these generated items -- I think Lehigh did that until recently in a similar scenario.  One option would be to just extract the entire contents of the `if ($number == 0)` block into a helper method, for that to be overridden as needed.  But of course that could be a later enhancement and isn't necessary at this point.

**Review by maccabeelevine — COMMENTED** [review]
> One more question: should we add a Folio.ini setting to toggle this behavior, in case some people actually want the legacy behavior?
> Once that is settled, perhaps we should solicit a couple more reviews from the FOLIO community to be sure nothing has been overlooked, and then I think this can be merged!

Yeah I think this is worth some discussion maybe in #vufind-folio-integrations.  For example the holdings-without-items pattern is common (I think) for electronic holdings, and holdings statements in particular for ejournals.  I'm not sure what all the side effects could be, but there may be reasons to disable it.

I'd also like a few more days myself, we're doing some related testing at Lehigh.

File: module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php]: Now that you mention it, @maccabeelevine, I think there's an important detail missing: this property can be set, but it's not actually used for anything. I think the isVisibleInHoldings() method needs adjustment too, and if that's the case, giving the property a more specific name definitely makes sense.

**Review by demiankatz — CHANGES_REQUESTED** [review]

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Good friendly amendment...I'll update the comment.

File: module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php
Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php]: Oh, nuts — that was sloppy of me. Yes, there is [currently code in AvailabilityStatus](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php#L89-L98) that I forgot to change to make `setVisibility()` actually have an effect. Will fix.

If we decide to change the name of this method, I think the only place it is used is in [View/Helper/Root/Holdings::holdingIsVisible()](https://github.com/vufind-org/vufind/blame/dev/module/VuFind/src/VuFind/View/Helper/Root/Holdings.php#L75).

Lastly, I went with `$visibilityInHoldingsTab` as a more descriptive name.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php:886]: I haven't been deep enough into VuFind to understand the inconsistency...which is to say, I'm not sure what kind of "format" action is taking place...it looks like it is filling the contents of VuFind's item array with the response from FOLIO's API, and this refactoring isn't substantially different.

I thought about pulling out the contents of `if ($number == 0)` into a separate method, but I couldn't think of any use cases where that would be useful. Happy to make the change if y'all thing it would be wise.

Comment (maccabeelevine) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php:886]: I'd say, ignore this suggestion and continue as-is.  If I come up with a concrete use case later on, it's an easy enough refactoring to do then.

File: module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php]: Looks much better to me, though I haven't re-tested yet (I'm holding off on doing another round of testing on my end until @maccabeelevine's testing is wrapped up and we've reached a conclusion about whether to make the behavior configurable... that way I'll keep my eyes fresher for the (hopefully) final round of testing). :-)

Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php]: Thanks, Demian...sounds like a good plan to me.

**Review by maccabeelevine — CHANGES_REQUESTED** [review]
Code looks fine to me now, but something is missing.  I'm testing on an ebook record (one holding, zero items) and I see this.  On `dev` there would be the Internet row but no Electronic holdings listing.  Does that have to be avoided in holdingsils/electronic.phtml?

image

**Comment by demiankatz** [comment]
@maccabeelevine, I thought the whole point of this PR was to cause the holding-without-items to be displayed in the holdings tab, which appears to be what is happening. But maybe I misunderstand the problem you are describing. Note also that the hide_holdings[] setting in config.ini can be used to suppress specific locations, which might be useful in combination with the new behavior if we actually do not want to display some empty locations but we do want to see others. (I still think making this configurable in Folio.ini would be wise -- but if we do, a note about hide_holdings in the comments would likely be helpful for supporting "mixed" use cases).

**Comment by maccabeelevine** [comment]
> @maccabeelevine, I thought the whole point of this PR was to cause the holding-without-items to be displayed in the holdings tab, which appears to be what is happening. But maybe I misunderstand the problem you are describing. Note also that the hide_holdings[] setting in config.ini can be used to suppress specific locations, which might be useful in combination with the new behavior if we actually do not want to display some empty locations but we do want to see others. (I still think making this configurable in Folio.ini would be wise -- but if we do, a note about hide_holdings in the comments would likely be helpful for supporting "mixed" use cases).

Ok, you're right, initially I thought the point was to display it in the search results, not in the holdings tab, but the screenshot in Jira shows both.  Given that, I agree that we need a configuration toggle in Folio.ini, or as you say we can just use the hide_holdings[] setting.

File: module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Logic/AvailabilityStatus.php]: @dltj, sounds like @maccabeelevine agrees that this should be configurable. Is that on your radar at this point? If you need any help, let me know!

File: config/vufind/Folio.ini
Comment (demiankatz) [inline config/vufind/Folio.ini]: Would it be safer to default to false to prevent unexpected changes in behavior? Or do you think the benefit is great enough that we should change the default behavior? (At Villanova, I'm pretty sure we _don't_ want this, so I'm possibly biased ;-) ).

Comment (dltj) [reply config/vufind/Folio.ini]: Hmmm... I was judging by the number of libraries that were asking Index Data for it, but I don't know how biased that sample is. Happy to change the default if desired.

Comment (maccabeelevine) [reply config/vufind/Folio.ini]: 6m ago I would have locally enabled this, now after some data changes I would locally disable.  The default is a hard call, this is very useful behavior when you have a record structured this way, but it's also not necessary intuitive, and something that a library should maybe consider when setting up FOLIO integration based on their own holdings.  So I guess I would default to false but I don't feel strongly about it.

Comment (demiankatz) [reply config/vufind/Folio.ini]: Maybe the way to go is false for now (if we want this in release 10.1) for a more backward-compatible approach, and we can revisit the default in the future if demand seems to push us that way.

My main concern if we change the default is that we'll get an influx of support requests about "why are these phantom items showing up?" Since this feature generates "fake" data (at least according to VuFind's model of things), it makes more sense to me as an opt-in, where people turn it on after they understand what it is for, rather than having to figure out the behavior when they unexpectedly encounter it.

But I'm still open to counter-arguments if anyone feels strongly the other way.

Comment (dltj) [reply config/vufind/Folio.ini]: Good enough reasoning for me. New merge coming.

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/Folio.php]: We need the $locationName variable since it is used in multiple places below, but would it be more concise to directly reference $locationData below instead of creating $locationCode and $locationIsActive, since they're never reused and don't really make a big difference to readability?

**Review by demiankatz — APPROVED** [review]
Thanks, @dltj! I think this is good to merge now, though I spotted one possible minor optimization which I'm highlighting below. If you want to make the change, great; if you don't think it's a good idea, feel free to ignore.

Either way, since @maccabeelevine also has a "requests changes" review, we'll need him to re-review this before it can be merged.

Thank you both!

**Review by maccabeelevine — APPROVED** [review]
Other than Demian's suggestion about location variables, this LGTM!

**Comment by demiankatz** [comment]
Thanks, @maccabeelevine -- so @dltj, please let me know how you'd like to proceed. If you think my suggestion is worthwhile but don't have time, I'm also happy to make the edit on your behalf if that would help. :-)

File: module/VuFind/src/VuFind/ILS/Driver/Folio.php
Comment (dltj) [reply module/VuFind/src/VuFind/ILS/Driver/Folio.php]: Minor matter of style, and I'm happy with either direction. I removed those under-utilized variables.

**Review by demiankatz — APPROVED** [review]
Thanks again, @dltj -- I know quite a few people will be happy to have this one!

**Comment by dltj** [comment]
Woo, hoo! Thanks Demian, Thomas, and Maccabee for your help!

**Pull request merged**
